fe/fe-core/src/main/java/com/starrocks/planner/PlanNode.java

```diff
@@ -710,16 +712,132 @@ public boolean canPushDownRuntimeFilter() {
+     */
+    public List<Expr> candidatesOfSlotExpr(Expr expr) {
+        List<Expr> candidateSlotExprs = Lists.newArrayList();
+        // NOTE: No need to check expr is slot or not here, each node should implement its `candidatesOfSlotExpr` itself.
+        if (!expr.isBoundByTupleIds(getTupleIds())) {
+            return candidateSlotExprs;
+        }
+        // only contain the input expr by default.
+        candidateSlotExprs.add(expr);
+        return candidateSlotExprs;
+    }
+
+    public List<List<Expr>> candidatesOfSlotExprs(List<Expr> exprs) {
+        List<List<Expr>> candidatesOfSlotExprs = Lists.newArrayList();
+        for (Expr expr: exprs) {
+            List<Expr> candidates = candidatesOfSlotExpr(expr);
+            if (candidates.isEmpty()) {
+                return Lists.newArrayList();
+            }
+            candidatesOfSlotExprs.add(candidates);
+        }
+        return permutaionsOfPartitionByExprs(candidatesOfSlotExprs);
+    }
+
+    public static List<List<Expr>> permutaionsOfPartitionByExprs(List<List<Expr>> partitionByExprs) {
+        if (partitionByExprs.isEmpty()) {
+            return Lists.newArrayList();
+        }
+
+        int totalCount = 1;
+        for (List<Expr> exprs: partitionByExprs) {
+            totalCount *= exprs.size();
+        }
+
+        List<List<Expr>> permutations = Lists.newArrayList();
+        boolean isGreaterThanExpect = totalCount > 16;
+        if (isGreaterThanExpect) {
+            totalCount = 16;
+        }
+
+        if (totalCount <= 0) {
+            return Lists.newArrayList();
+        }
+
+        for (int i = 0; i < totalCount; i++) {
+            permutations.add(Lists.newArrayList());
+        }
+
+        for (int k = 0; k < partitionByExprs.size() - 1; k++) {
+            List<Expr> exprs = partitionByExprs.get(k);
+            int numTimes = totalCount / exprs.size();
+            for (int i = 0; i < exprs.size(); i++) {
+                for (int j = 0; j < numTimes; j++) {
+                    if ((i * numTimes) + j < totalCount) {
+                        permutations.get((i * numTimes) + j).add(exprs.get(i));
+                    }
+                }
+            }
+        }
+        // output the last column
+        List<Expr> exprs = partitionByExprs.get(partitionByExprs.size() - 1);
+        int numTimes = totalCount / exprs.size();
+        for (int i = 0; i < exprs.size(); i++) {
+            for (int j = 0; j < numTimes; j++) {
+                if (i + (j * numTimes) < totalCount) {
+                    permutations.get(i + (j * numTimes)).add(exprs.get(i));
+                }
+            }
+        }
+        if (isGreaterThanExpect) {
+            return permutations.stream().filter(perm -> perm.size() == partitionByExprs.size()).collect(Collectors.toList());
+        }
+        return permutations;
+    }
+
+    public Pair<Boolean, List<List<Expr>>> canPushDownRuntimeFilterCrossExchange(
```

Comment:
suggest to define a class as return value, doing so can make the code easy to understand.

class RuntimeFilterPushDownResult  {
public boolean pushdown;
public Optional<List<Expr>> partitionByExprs;
}